Search backend implementation

LotteHofstede opened the pull request:
Worked on using the REST API's search endpoint instead of the mock search service. 
Replaced the `search` method in `search.service.ts` with a real implementation.

Refactored parts of the data builders, to make sure the type of the requested item does not need to be known and can be retrieved from the object's type.
Made sure the right SearchResultObjects are created, based on the DSpaceObject they wrap.

At the moment 5 tests still fail, we are aware of this: this is still a work in progress.

artlowel commented:
This PR closes #135

atarix83 commented:
This is a first review after using part of the actually Search implementation for the mydspace page :
 * because of we're using in mydspace implementation I should move search service and all components related, to the `shared` directory
 * I should change the uiSearchRoute param in `SearchService` according to the ui route, as we made [here ](https://github.com/4Science/dspace-angular/blob/6311884665a9830831d69cfb2bb3e2e19ac722fe/src/app/%2Bsearch-page/search-service/search.service.ts#L300)
 * In mydspace we are using the param `configuration` of the Search rest endpoint, 
   so we need to pass it to SearchService. I think it should be a property of the [`SearchOptions`](https://github.com/LotteHofstede/dspace-angular/blob/search-backend/src/app/%2Bsearch-page/search-options.model.ts) class
 * In the mock version of the search module, loading spinner was shown during the search. Now it doesn't happen but i think it's useful
 * another improvement should be to show a message whether the search does not return any results.

 * When the search doesn't return any reults this error occurs :
  ``` ERROR TypeError: Cannot read property 'objects' of undefined
    at MapSubscriber.project (search.service.ts:134)
    at MapSubscriber._next (map.js:79)
    at MapSubscriber.Subscriber.next (Subscriber.js:90)
    at MapSubscriber._next (map.js:85)
    at MapSubscriber.Subscriber.next (Subscriber.js:90)
    at MapSubscriber._next (map.js:85)
    at MapSubscriber.Subscriber.next (Subscriber.js:90)
    at MergeMapSubscriber.notifyNext (mergeMap.js:145)
    at InnerSubscriber._next (InnerSubscriber.js:23)
    at InnerSubscriber.Subscriber.next (Subscriber.js:90)
 ```
We have fixed adding a better filter [here ](https://github.com/LotteHofstede/dspace-angular/blob/86fcf44977601db5efc86b50de3af17ecc1374e5/src/app/core/data/search-response-parsing.service.ts#L34) and [here ](https://github.com/LotteHofstede/dspace-angular/blob/86fcf44977601db5efc86b50de3af17ecc1374e5/src/app/core/data/search-response-parsing.service.ts#L48). You can view changes [here ](https://github.com/4Science/dspace-angular/blob/dspacecris_demo/src/app/core/data/search-response-parsing.service.ts)

    
  * While implementig the mydspace page to retrieve filters config and facet values,
    we have used only the response from `/api/discover/search/objects` endpoint.
    This should save other further REST call to build search results.
    If this could be useful the full code is available [here](https://github.com/4Science/dspace-angular/tree/dspacecris_demo)

LotteHofstede commented:
@atarix83 Thanks for your feedback, I made some new commits to solve some of these issues.

> because of we're using in mydspace implementation I should move search service and all components related, to the shared directory

I don't think this is necessary. You can simply import the search module into the mydspace module if you need it there. If we move this to the `shared` folder, everything might end up in there and the `shared` folder will lose its purpose.

> I should change the uiSearchRoute param in SearchService according to the ui route, as we made here

You're right. I changed this to make sure the current route is used instead of the hardcoded `"/search"` path. [I reused your code for this.](https://github.com/LotteHofstede/dspace-angular/commit/e56ea2aff894ea21537d66340b74b8a7b6a59787#diff-4ef3e816a518eb897da43f7f538cf230R238)

> In mydspace we are using the param configuration of the Search rest endpoint,
so we need to pass it to SearchService. I think it should be a property of the SearchOptions class

I believe I already did that, somewhere in the process. Both the sorting and pagination parameters as well as the current filters, are saved in the[ (Paginated)SearchOptions](https://github.com/LotteHofstede/dspace-angular/blob/search-backend/src/app/+search-page/paginated-search-options.model.ts) object now.

> In the mock version of the search module, loading spinner was shown during the search. Now it doesn't happen but i think it's useful

I made sure the loader component is also shown when the [result object does not exist](https://github.com/LotteHofstede/dspace-angular/blob/search-backend/src/app/+search-page/search-results/search-results.component.html#L11) yet. I saw this was already done in the collection-page component, but not in any other components. This might explain why the spinner doesn't appear on the [home page](https://dspace7-demo.atmire.com/home) either.

> another improvement should be to show a message whether the search does not return any results.

I added a message when the result set is an empty array.

> When the search doesn't return any reults this error occurs

I could not reproduce this error (anymore). I did solve another issue where the result list would not exist if there were no results. This didn't cause any errors in my console, but I wasn't able to determine when the request was finished or not, without fixing this first.

> While implementig the mydspace page to retrieve filters config and facet values,
we have used only the response from /api/discover/search/objects endpoint.
This should save other further REST call to build search results.
If this could be useful the full code is available here 

In my opinion, it should not matter what endpoint is called. What I did is call the specific endpoints I need. If this data was already returned by another request, it should have been cached. I know it wasn't implemented that way yet, but I don't think this should be the `SearchService`'s concern. If you would want to show the facets on another page, you would still have to ask the `SearchService` for this data. It would not make sense to request the whole `/objects` endpoint in this case.

PTrottier commented on karma.conf.js:
```diff
@@ -51,7 +51,7 @@ module.exports = function (config) {
     files: [{
       pattern: './spec-bundle.js',
-      watched: false
+      watched: false,
```
I often omit the trailing comma for the last element.

What's the reason behind this change?

PTrottier replied:
I often omit the trailing comma for the last element.

What's the reason behind this change?

PTrottier commented on src/app/+search-page/search-service/search.service.spec.ts:
```diff
@@ -8,50 +8,242 @@ import { SearchService } from './search.service';
+        imports: [
+          CommonModule,
+          RouterTestingModule.withRoutes([
+            { path: 'search', component: DummyComponent, pathMatch: 'full' },
```
See the previous comment.

PTrottier commented on src/app/+search-page/search-filters/search-filter/search-facet-filter/search-facet-filter.component.html:
```diff
@@ -2,34 +2,37 @@
+               [placeholder]="'search.filters.filter.' + filterConfig.name + '.placeholder'| translate"  [ngModelOptions]="{standalone: true}"/>
         <input type="submit" class="d-none"/>
     </form>
-</div>
```
Should there be a trailing return case at the end of the file?

artlowel commented:
@tdonohue:

>3. If I visit a Collection Homepage, the "Recent Submissions" never loads.  In the Chrome Developer Tools Console, I see this error message: `Error: A browse endpoint for items on dc.date.issued isn't configured`.  (I tested this on the latest `master` and it works perfectly there, so this error seems to be caused by this PR)

Fixed by c125f36f43403c320258bb0ef4571ee3500770bc, forgot `this.` in a refactor

>4. In the Search filters, the "Date" filter doesn't work (it lists possible values, but selecting one always returns "no results"). All other filters function properly.

This will be fixed in the PR that adds the proper date widget as described in the mockups.

>5. "Sort By" option in Settings doesn't seem to affect the results listing. It changes the URL in the Chrome address bar (e.g. `sortDirection=DESC`), but the results listing remain in the same order.

Fixed. Reason was a peculiarity with typescript string enums. Apparently you can't fetch a key by its value (only for strings, it works for other enums). This was fixed by making keys and values identical.

>6. "Results per page" option in Settings works fine in the list view. However, it doesn't seem to function properly in the grid view. If I select a different option, it always resets itself back to whatever value it last was in the list view. For example, if I switch to "list view" and select "5" results per page, this results in multiple pages.  If I then switch to "grid view" and select a higher value (e.g. "12"), pagination remains.

Lotte ran in to some issues with the different page sizes for grids and lists. Since you already noted that it wasn't very clear to an end user what was going on, we decided to scrap the idea, but there were some remnants left over. I've removed those now, and both views, now use the same page sizes again.

@PTrottier:
These will probably have happened automatically by the IDE. [Trailing commas](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas) are no problem in JS, some people even encourage their use, as it means you can add a next property without having to modify the previous line, so you get a cleaner diff in git.

A missing newline at the end of a file _can_ cause issues with certain tools, so I've added it.

tdonohue commented:
I've re-tested this PR today and all the bugs I'd previously reported (other than the date filter, as noted) were fixed. This works great now, so I've given it a 👍 .

However, I will note one other oddity I just encountered. It's not really a "bug", but it's definitely unexpected behavior. If I use the demo REST API with this PR and search on "standards-based metadata procedure", one might expect that the *top* match would be the document with this title:

"Standards-based metadata procedures for retrieving data for display or mining utilizing persistent (data-DOI) identifiers"

But, it's **not** the top match. Instead, the results come back **ordered alphabetically**, so that the "best match" is listed third (out of three). The other two matches also do contain these words (in either title or abstract), but not all words even appear in the title. Here's the three titles in the order they are returned:

1. "Science, institutional archives and open access: an overview and a pilot survey on the Italian cancer research institutions" (Zero matches in title, but all words are in abstract)
2. "Standards-based curation of a decade-old digital repository dataset of molecular information" (One word in title, and others are in abstract)
3. "Standards-based metadata procedures for retrieving data for display or mining utilizing persistent (data-DOI) identifiers" (This is the document I'd expect to see *first* as what I searched on is the beginning of this title)

LotteHofstede commented:
@tdonohue The default sort order was set to `dc.title` (ascending) in the `SortOptions` constructor. I removed the default parameters, and made sure that it's set by the responsible components and services. The search page is now sorted by `score` (descending) by default. The other components still order by `dc.title` (ascending).
